numpy/core/src/multiarray/item_selection.c

```diff
@@ -1292,7 +1292,14 @@ partition_prep_kth_array(PyArrayObject * ktharray,
     npy_intp * kth;
     npy_intp nkth, i;
 
-    if (!PyArray_CanCastSafely(PyArray_TYPE(ktharray), NPY_INTP)) {
+    if (PyArray_ISBOOL(ktharray)) {
+        /* 2021-09-29, NumPy 1.22 */
+        if (DEPRECATE(
+                "Passing booleans as partition index is deprecated") < 0) {
+            return NULL;
+        }
+    }
+    else if (!PyArray_ISINTEGER(ktharray)) {
```

Comment:
OK, I think we may have to switch hat over to a same-kind casting (possibly with a time exception).  But it is what indexing currently uses, so I think it is fine.  (This would not allow custom user integer DTypes.)

Reply:
For future refernce: I wonder what would be the best way of dealing with custom user integer DTypes?
Considering `PyArray_ISINTEGER` is too strict and `np.can_cast(..., casting="same_kind")` won't even accept all the standard numpy integer dtypes.

Reply:
> All NumPy integers should pass fine?

Ah, to clarify: I was talking specifically about `np.uint64` in the `can_cast` example.

> If same-kind casting doesn't do the trick, we could also consider adding an abstract `Indexer` DType that users have to register or implement promotion with.

I could see this potentially being usefull, as this would also make it easier to differentiate between (user-defined) bool dtypes.

Reply:
Ugh, you're completelly right.
I may have accidently mixed up "safe" and "same_kind" in my head...